File: 210-course-schedule-ii/course-schedule-ii.py
class Solution:
    def findOrder(self, n: int, arr: List[List[int]]) -> List[int]:
        adj = {}
        for i in range(n):
            adj[i] = []
        for u,v in arr:
            adj[u].append(v)
        visited = set()
        
        def dfs(node):
            nonlocal res
            if node in visited:
                return True
            if node in seen:
                return False
            # if node in cache:
            #     return list(cache[node])
            seen.add(node)
            for prereq in adj[node]:
                if not dfs(prereq):
                    return False
                # print('node,prereq,curr',node,prereq)
            if node not in visited:
                res.append(node)
                visited.add(node)
            return True
            

        res = []
        for i in range(n):
            seen = set()
            if i not in visited:
                if not dfs(i):
                    return []
        return res

File: 210-course-schedule-ii/course-schedule-ii.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function(n, prereq) {
    const indegree = Array(n).fill(0)
    const adj = {}
    for(let i=0;i<n;i++){
        adj[i] = []
    }
    for(const [u,v] of prereq){
        adj[v].push(u)
        indegree[u]+=1
    }
    const q = []
    const res = []
    for(let i=0;i<n;i++){
        if(indegree[i]==0){
            q.push(i)
        }
    }
    while(q.length>0){
        const node = q.shift()
        res.push(node)
        for(let nei of adj[node]){
            indegree[nei]-=1
            if (indegree[nei]==0){
                q.push(nei)
            }
        }
    }
    return res.length == n? res:[]

};